File: solutions/java/prism/1/src/main/java/Prism.java
import java.util.*;

public class Prism {
    private static final double MAX_DISTANCE_FROM_RAY = 0.11;
    private static final double FORWARD_EPS = 1e-9;

    public record LaserInfo(double x, double y, double angle) {
    }

    public record PrismInfo(int id, double x, double y, double angle) {
    }

    public static List<Integer> findSequence(LaserInfo laser, List<PrismInfo> prisms) {
        List<Integer> answer = new ArrayList<>();

        double currentX = laser.x();
        double currentY = laser.y();
        double currentAngle = laser.angle();

        while (true) {
            double angle = normalizeAngle(currentAngle);
            double radians = Math.toRadians(angle);

            double dx = Math.cos(radians);
            double dy = Math.sin(radians);

            PrismInfo nextPrism = null;
            double bestDistance = Double.POSITIVE_INFINITY;

            for (PrismInfo prism : prisms) {
                double vx = prism.x() - currentX;
                double vy = prism.y() - currentY;

                double cross = dx * vy - dy * vx;
                if (Math.abs(cross) > MAX_DISTANCE_FROM_RAY) {
                    continue;
                }

                double dot = dx * vx + dy * vy;
                if (dot <= FORWARD_EPS) {
                    continue;
                }

                if (dot < bestDistance) {
                    bestDistance = dot;
                    nextPrism = prism;
                }
            }

            if (nextPrism == null) {
                break;
            }

            answer.add(nextPrism.id());
            currentX = nextPrism.x();
            currentY = nextPrism.y();
            currentAngle += nextPrism.angle();
        }

        return answer;
    }

    private static double normalizeAngle(double angle) {
        double normalized = angle % 360.0;
        if (normalized < 0) {
            normalized += 360.0;
        }
        return normalized;
    }
}

File: solutions/csharp/prism/1/Prism.cs
public static class Prism
{
    public readonly record struct LaserInfo(double X, double Y, double Angle);

    public readonly record struct PrismInfo(int Id, double X, double Y, double Angle);

    public static int[] FindSequence(LaserInfo laser, PrismInfo[] prisms)
    {
        var result = new List<int>();

        double x = laser.X;
        double y = laser.Y;
        double angle = laser.Angle;

        int maxSteps = prisms.Length * 5;

        for (int step = 0; step < maxSteps; step++)
        {
            PrismInfo? nextPrism = null;
            double minDist = double.MaxValue;

            double rad = angle * Math.PI / 180.0;
            double dx = Math.Cos(rad);
            double dy = Math.Sin(rad);

            foreach (var p in prisms)
            {
                double vx = p.X - x;
                double vy = p.Y - y;
            
                double dist = Math.Sqrt(vx * vx + vy * vy);
                if (dist == 0) continue;
            
                double dirX = vx / dist;
                double dirY = vy / dist;

                double dot = dirX * dx + dirY * dy;

                if (dot <= 0.999) continue; 
            
                if (dist < minDist)
                {
                    minDist = dist;
                    nextPrism = p;
                }
            }

            if (nextPrism == null)
            {
                break;
            }

            var hit = nextPrism.Value;
            result.Add(hit.Id);

            x = hit.X;
            y = hit.Y;

            angle += hit.Angle;

            if (angle >= 360) angle %= 360;
            if (angle < 0) angle = 360 + (angle % 360);
        }
        return result.ToArray();
    }
}